### ISSUE TITLE
Bug: Cannot install on Fedora Silverblue or IoT

### ISSUE THREAD

**Issue by daniel-wtd** [performancecopilot/pcp#816]
Hi,

i am unable to perform an installation in Fedora IoT (31) and Fedora Silverblue (31). The package installation is progressing ok, until the "post scripts":

```
Processing packages... done
Running pre scripts... done
Running post scripts... done
error: Running %post for pcp: Executing bwrap(/bin/sh): Child process killed by signal 1; run `journalctl -t 'rpm-ostree(pcp.post)'` for more information
```

Thank you so much in advance for having a look. If I can support you in any way, please feel free to reach out to me.

**Comment by natoscott** [comment]
I think I know what this one is, but are there any extra diagnostics from ...

| run `journalctl -t 'rpm-ostree(pcp.post)'` for more information

thanks.

**Comment by daniel-wtd** [comment]
Hi,

thank you for having a look.

It seems to be related to "creation of files and directories in post".

```
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: touch: cannot touch '/var/lib/pcp/pmns/.NeedRebuild': No such file or directory
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: chmod: cannot access '/var/lib/pcp/pmns/.NeedRebuild': No such file or directory
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: rpm-ostree-systemctl: Ignored non-preset command: try-restart pmcd.service
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: Created symlink /etc/systemd/system/multi-user.target.wants/pmcd.service → /usr/lib/systemd/system/pmcd.service.
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: rpm-ostree-systemctl: Ignored non-preset command: try-restart pmlogger.service
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: Created symlink /etc/systemd/system/multi-user.target.wants/pmlogger.service → /usr/lib/systemd/system/pmlogger.service.
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: rpm-ostree-systemctl: Ignored non-preset command: try-restart pmie.service
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: Created symlink /etc/systemd/system/multi-user.target.wants/pmie.service → /usr/lib/systemd/system/pmie.service.
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: /usr/pcp.post: line 53: cd: /var/lib/pcp/pmns: No such file or directory
Jan 08 22:29:02 nb01 rpm-ostree(pcp.post)[2449]: /usr/pcp.post: line 54: cd: HOME not set
```

**Comment by natoscott** [comment]
@daniel-wtd yeah, that's what I was expecting. :(

I think these specific issues (relating to creating the metric namespace during %post) could be fixed by changing the rpm spec file to:

- install an empty %file for .NeedRebuild - will need to be a %ghost as it gets deleted - instead of using touch(1)
- not fail the rpm install if ./Rebuild fails (e.g. if /var/lib/pcp/pmns is read-only as it seems to be here)

We currently run the metric namespace (pmns) ./Rebuild in %post because use of local-context mode (e.g. in dstat, and other tools) without pmcd means there is no other opportunity to do it (AFAICS) - it has to run as root, which these tools do not.

But, there may then be other, even more difficult to resolve issues with Silverblue - we'd need someone with spec file knowledge to try an install & see how it all goes.

**Comment by natoscott** [comment]
@daniel-wtd I'll push some changes into the PCP master branch shortly that will avoid with the initial issues you've observed here.  There may be other subsequent problems however, so could you then attempt another installation?  (./Makepkgs in the PCP git repo can be used to build new RPMs).  Thanks!

**Comment by daniel-wtd** [comment]
Sure, I will have a look these days :)

**Comment by daniel-wtd** [comment]
After installing the build dependencies I am still getting some errors during the tar build, that man page packages are missing :/ Any advice how to proceed here?

**Comment by natoscott** [comment]
@daniel-wtd can you add a link to a [f]paste / gist of the errors here?  thanks.

**Comment by daniel-wtd** [comment]
Sorry for the late response. :(

I have done the following:

```
toolbox create -c pcp
toolbox enter -c pcp
git clone https://github.com/performancecopilot/pcp.git
sudo dnf -y install gcc make flex bison perl-ExtUtils-MakeMaker systemd-devel
./Makepkgs --verbose
```
This resulted in

```
=== tar ===
make[1]: Entering directory '/var/home/dschier/Projects/temp/pcp/pcp-5.1.0/build/tar'
/usr/bin/gtar: Removing leading `etc/pcp/../' from member names
/usr/bin/gtar: usr/share/man/man3/PCP\:\:LogImport.3pm.gz: Cannot stat: No such file or directory
/usr/bin/gtar: usr/share/man/man3/PCP\:\:LogSummary.3pm.gz: Cannot stat: No such file or directory
/usr/bin/gtar: usr/share/man/man3/PCP\:\:MMV.3pm.gz: Cannot stat: No such file or directory
/usr/bin/gtar: usr/share/man/man3/PCP\:\:PMDA.3pm.gz: Cannot stat: No such file or directory
/usr/bin/gtar: Removing leading `usr/share/pcp/../' from member names
listfiles: 6226 files from the main manifest
listfiles: 14 Perl files
listfiles: 50 Python files
/usr/bin/gtar: Exiting with failure status due to previous errors
Wrote: /var/home/dschier/Projects/temp/pcp/pcp-5.1.0/build/tar/pcp-5.1.0-1.tar.gz
make[1]: Leaving directory '/var/home/dschier/Projects/temp/pcp/pcp-5.1.0/build/tar'
Done
make: Leaving directory '/var/home/dschier/Projects/temp/pcp/pcp-5.1.0/build'

```

**Comment by natoscott** [comment]
@daniel-wtd I'm not sure what 'toolbox' is, but this looks like a build failure unrelated to Silverblue.  You should be able to just run the build without perl (--without-perl), or peruse the output from ./qa/admin/check-vm for missing build machine packages.

**Comment by daniel-wtd** [comment]
without perl, the build was going through, but no rpm or deb was built.

I tested the qa-script which was flodding me with a ton of stuff ^^ I will try t resolve everything and come back.

```
./qa/admin/check-vm 
Missing: ed [ed]
Missing: gdb [gdb]
Missing: autoconf [autoconf]
Missing: rpmbuild [rpm-build]
Missing: cpan [perl-CPAN or perl]
Missing: mandoc [mandoc (build optional)]
Missing: cppcheck [cppcheck (build optional)]
Missing: /usr/include/uv.h [libuv-devel (build optional)]
Missing: /usr/include/varlink.h [libvarlink-devel (build optional)]
Missing: /usr/include/lzma.h [xz-devel (build optional)]
Missing: /usr/include/sasl/sasl.h [cyrus-sasl-devel (build optional)]
Missing: /usr/include/infiniband/umad.h [rdma-core-devel or libibumad-devel (build optional)]
Missing: /usr/include/infiniband/mad.h [infiniband-diags-devel or libibmad-devel (build optional)]
Missing: /usr/include/avahi-common/defs.h [avahi-devel or libavahi-devel (build optional)]
Missing: /usr/include/perfmon/pfmlib_perf_event.h [libpfm-devel (build optional)]
Missing: /usr/include/rpm/header.h [rpm-devel (build optional)]
Missing: g++ [gcc-c++ or ccache]
Missing: /usr/include/readline/readline.h [readline-devel]
Missing: /usr/include/ncurses.h [ncurses-devel]
Missing: /usr/lib*/python3.*/*-packages/jsonpointer.py [python3-jsonpointer]
Missing: /usr/lib*/python3.*/*-packages/rtslib [python3-rtslib or python3-rtslib-fb]
Missing: /usr/lib*/python3.*/*-packages/libvirt.py [libvirt-python3 or python3-libvirt or python3-libvirt-python (build optional)]
Missing: /usr/lib*/python3.*/*-packages/lxml/sax.py [python3-lxml or python36-lxml (build optional)]
Missing: /usr/lib*/python3.*/*-packages/prometheus_client [python3-prometheus_client (build optional)]
Missing: /usr/lib*/python3.*/*-packages/elasticsearch [python3-elasticsearch (build optional)]
Missing: /usr/lib*/python3.*/*-packages/openpyxl [python3-openpyxl (build optional)]
Missing: /usr/lib/python3.*/*-packages/bcc/__init__.py [python3-bcc (build optional)]
Missing: bpftrace [bpftrace (build optional)]
Missing: /usr/lib*/python3.*/*-packages/PIL/Image.py [python3-pillow or python3-Pillow or python36-pillow (QA optional)]
Missing: /usr/lib*/python3.*/*-packages/psycopg2 [python3-psycopg2 (postgresql)]
Missing: /usr/lib*/python3.*/*-packages/pyodbc* [python3-pyodbc (mssql QA optional)]
Missing: Date::Format [perl-TimeDate or perl(Date::Format)]
Missing: Date::Manip [perl-Date-Manip or perl(Date::Manip)]
Missing: Date::Parse [perl-TimeDate or perl(Date::Parse)]
Missing: YAML::XS::LibYAML [perl-YAML-LibYAML or perl(YAML::LibYAML) or cpan(YAML::LibYAML)]
Missing: XML::LibXML [perl-XML-LibXML or perl(XML::LibXML) or cpan(XML::LibXML) (QA optional)]
Missing: XML::TokeParser [perl-XML-TokeParser or perl(XML::TokeParser) or cpan(XML::TokeParser)]
Missing: File::Slurp [perl-File-Slurp or perl(File::Slurp) (QA optional)]
Missing: List::MoreUtils [perl-List-MoreUtils or perl(List::MoreUtils) (QA optional)]
Missing: autodie:: [perl or perl-autodie or perl(autodie) (QA optional)]
Missing: RRDs:: [rrdtool-perl or perl-rrdtool or rrdtool or perl(RRDs)]
Missing: Spreadsheet::Read [perl-Spreadsheet-Read or perl(Spreadsheet::Read) or cpan(Spreadsheet::Read) (QA optional)]
Missing: Spreadsheet::WriteExcel [perl-Spreadsheet-WriteExcel or perl(Spreadsheet::WriteExcel) or cpan(Spreadsheet::WriteExcel)]
Missing: Spreadsheet::ReadSXC [perl-Spreadsheet-ReadSXC or perl(Spreadsheet::ReadSXC) or cpan(Spreadsheet::ReadSXC) (QA optional)]
Missing: Spreadsheet::XLSX [perl-Spreadsheet-XLSX or perl(Spreadsheet::XLSX) or cpan(Spreadsheet::XLSX) (QA optional)]
Missing: Text::CSV_XS [perl-Text-CSV_XS or perl(Text::CSV_XS) or cpan(Text::CSV_XS)] (QA optional)]
Missing: JSON:: [perl-JSON or perl(JSON)]
Missing: Net::LDAP [perl-LDAP or perl-ldap or perl(Net::LDAP)]
Missing: LWP::UserAgent [perl-libwww-perl or perl(LWP::UserAgent)]
Missing: Net::SNMP [perl-Net-SNMP or perl(Net::SNMP) or cpan(Net::SNMP)]
Missing: sadf [sysstat (for sar2pcp)]
Missing: host [bind-utils]
Missing: bc [bc (QA optional)]
Missing: socat [socat (QA optional)]
Missing: nmap [nmap (QA optional)]
Missing: valgrind [valgrind (QA optional)]
Missing: avahi-browse [avahi-tools or avahi-utils (QA optional)]
Missing: ifconfig [net-tools or net-tools-deprecated]
Missing: netstat [net-tools or net-tools-deprecated]
Missing: kubectl [kubernetes-client or kubectl]
Missing: kubelet [kubernetes-node]
Missing: crontab [cronie]
Missing: mail [mailx (QA optional)]
Missing: httpd|/etc/apache2/httpd.conf [httpd or apache2 (QA optional)]
Missing: /usr/lib*/python3.*/*-packages/requests [python3-requests (QA optional)]
Missing: /usr/bin/logconv.pl|/usr/lib/389-ds/bin/logconv.pl [389-ds-base or 389-ds (QA optional)]
Missing: /usr/lib*/sasl2/libplain.so [cyrus-sasl-plain (QA optional)]
Missing: /usr/lib*/libicui18n.so [libicu-devel (QA optional)]
Missing: DBD::mysql [perl-DBD-mysql or perl-DBD-MySQL or perl(DBD::mysql) or perl(DBD::MySQL) (mysql QA optional)]
Missing: memcached [memcached (memcache QA optional)]
Missing: zabbix_agentd [zabbix-agent or zabbix22-agent (zabbix QA optional)]
Missing: /opt/mssql/bin/sqlservr [mssql-server (mssql QA optional)]
Missing: /opt/mssql-tools/bin/sqlcmd [mssql-tools (mssql QA optional)]
Missing: psql [postgresql (postgresql QA optional)]
Missing: mkfs.gfs2 [gfs2-utils (gfs2 QA optional)]
Missing: docker [docker-common or docker or docker-io (QA optional)]
Missing: qtchooser|/usr/lib/qtchooser/qtchooser [qtchooser (build optional)]
Missing: qmake|/usr/lib*/qt4/bin/qmake [qt4-devel or libqt4-devel or qt-devel (build optional)]
Missing: /usr/lib*/libQtCore.so|/usr/lib/qt4/lib/libQtCore.so [qt4-devel or libqt4-devel or qt-devel (build optional)]
Missing: /usr/lib*/libQt5Core.so|/usr/lib/qt5/lib/libQtCore.so [qt5-qtbase-devel or libqt5Core-devel or libQt5Core-devel or libqt5-qtbase-devel (build optional)]
Missing: uic|/usr/lib/qtchooser/uic [qtchooser (build optional)]
Missing: /usr/include/qt5/QtSvg/QtSvg [qt5-qtsvg-devel or libqt5-qtsvg-devel (build optional)]
Missing: /usr/include/Inventor/Qt/SoQt.h|/usr/include/Coin*/Inventor/Qt/SoQt.h [SoQt-devel (build optional)]
Missing: /usr/include/Inventor/SoPath.h|/usr/include/Coin3/Inventor/SoPath.h [Coin-devel (build optional)]
Missing: /usr/lib*/libSoQt.so [SoQt-devel (build optional)]
Missing: /usr/include/nspr4/nspr.h [nspr-devel or mozilla-nspr-devel (build optional)]
Missing: /usr/include/nss3/nss.h [nss-devel or mozilla-nss-devel (build optional)]
Missing: certutil [nss-tools or mozilla-nss-tools (QA optional)]
Missing: /usr/include/openssl/ssl.h [openssl-devel (build optional)]
Missing: pluginviewer [cyrus-sasl (QA optional)]
Missing: seinfo [setools-console or setools (build optional)]
Missing: /usr/share/selinux/devel/Makefile [selinux-policy-devel or selinux-policy (build optional)]
Missing: /etc/selinux/targeted/policy [selinux-policy-targeted (build optional)]
Missing: smartctl [smartmontools (QA optional)]
Missing: ausyscall [audit (QA optional)]
Missing: sensors [lm_sensors or sensors (QA optional)]
Missing: mkfs.xfs [xfsprogs (QA optional)]
Missing: redis-server [redis (QA optional)]
Missing: zpool [zfs-fuse (QA optional)]
Missing: libvirtd [libvirt-daemon (QA optional)]
Missing: /usr/lib*/libvirt/*/libvirt_driver_qemu.so [libvirt-daemon-driver-qemu (QA optional)]
Missing: /usr/include/libdevmapper.h [device-mapper-devel (build optional)]
Missing: ruby [ruby (QA optional)]
Missing: statsd [statsd (QA optional)]
Missing: /usr/include/chan/chan.h [N/A on Fedora 31]
Missing: /usr/include/hdr/hdr_thread.h [N/A on Fedora 31]
Missing: /usr/sbin/qshape [postfix-perl-scripts]
Missing: /usr/sbin/service|/sbin/service [initscripts]
Warning: qt-devel not installed, need >= 4.4
Warning: qt4-devel not installed, need >= 4.4
Warning: libqt4-devel not installed, need >= 4.4
Warning: lib64qt4-devel not installed, need >= 4.4
Warning: libpfm-devel not installed, need >= 4.4 for pcp-pmda-perfevent
Warning: libpfm not installed, need >= 4.4 for pcp-pmda-perfevent
Warning: libpfm4 not installed, need >= 4.4 for pcp-pmda-perfevent
Warning: libibmad-devel not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: infiniband-diags-devel not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: libibmad not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: libibmad5 not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: infiniband-diags not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: libibumad-devel not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: rdma-core-devel not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: libibumad not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: libibumad3 not installed, need >= 1.1.7 for pcp-pmda-infiniband
Warning: Package libuv not known to pkg-config , need >= 1.0
No /etc/hosts entry for toolbox
Warning: "/etc/pcp.conf" is missing
sudo: unknown user: pcp
sudo: unable to initialize policy plugin
```

**Comment by natoscott** [comment]
@daniel-wtd could you try once more with the just released pcp-5.0.2?  Thanks!

**Comment by dnoliver** [comment]
I have tested the update, installation is successful, but services cannot start, logs are present here https://bugzilla.redhat.com/show_bug.cgi?id=1827441#c2

**Comment by daniel-wtd** [comment]
Thanks a lot for your time invested here. I really appreciate this.

Meanwhile, I have updated to F32. Nevertheless, using the testing repository, I was able to install pcp-5.1.0-1

After facing the same issues as @dnoliver , I was digging just a little bit more:

```
## lets assume selinux isnt an issue for now
setenforce permissive
```

```
## using regular systemd stuff
systemctl start pmcd
## its failed, lets check
systemctl status pmcd
journalctl -u pmcd
## nothing useful here, but we know pmie is needed for pmcd
systemct stats pmie
journalctl -u pmie
## as expected, this was one is not working either
```
Useful logs

```
## log
Apr 30 19:57:33 core00 systemd[1]: pmie.service: Can't open PID file /run/pcp/pmie.pid (yet?) after start: Operation not permitted
```
After some digging via bash -x for the scripts and checking whats going on, the issue seems to be related to:

```
# bash -x /usr/share/pcp/lib/pmcd start

+ /usr/libexec/pcp/bin/pmcd_wait
+ status=2
+ message='pmcd_wait failed in /usr/share/pcp/lib/pmcd: exit status: 2'
+ logger -p daemon.warning 'pmcd_wait failed in /usr/share/pcp/lib/pmcd: exit status: 2'
+ /usr/libexec/pcp/bin/pmpost 'pmcd_wait failed in /usr/share/pcp/lib/pmcd: exit status: 2'
+ _RC_STATUS -v
+ _rc_status=0
+ '[' -v = -v ']'
+ '[' 0 -eq 0 ']'
+ /usr/bin/echo
```

Checking said pmie executable does not help. Nevertheless, additional logs are made in /var/log/pcp.

```
# /var/log/pcp/pmcd/pmcd.log

Log for pmcd on core00 started Thu Apr 30 20:26:40 2020

Warning: localhost only from PMCD_LOCAL=1 in environment
Error Parsing ASCII PMNS: Cannot open "/var/lib/pcp/pmns/root"
Error: pmLoadASCIINameSpace(DEFAULT, 1): No such file or directory
[Thu Apr 30 20:26:40] pmcd(15469) Error: pmcd not started due to errors!

Log finished Thu Apr 30 20:26:40 2020
```

```
# /var/log/pmie/core00/pmie.log
Log for pmie on core00 started Thu Apr 30 19:58:00 2020

pmie: PID = 11011, via local:

pmie: cannot access config file as either config.default or /var/lib/pcp/config/pmie/config.default: No such file or directory

Log finished Thu Apr 30 19:58:00 2020
```

Creating these directories alone does not really improve the situation, even after changing them proper permissions. Some other namespace errors occur.

**Comment by goodwinos** [comment]
```
Error Parsing ASCII PMNS: Cannot open "/var/lib/pcp/pmns/root"
Error: pmLoadASCIINameSpace(DEFAULT, 1): No such file or directory
[Thu Apr 30 20:26:40] pmcd(15469) Error: pmcd not started due to errors!
```

Looks like the PMNS rebuild in %post scriptlet failed, presumably because the filesystem is read-only during the install. pmcd will not start if it cant load the PMNS root, as reported. To mitigate that, you could :

`sudo touch /var/lib/pcp/pmns/.NeedRebuild; sudo systemctl restart pmcd`

The suggestion to install /var/lib/pcp/pmcd/.NeedRebuild as a %ghost file (instead of touching it during the install %post scriptlet) should help, though there are other things probably still broken if the post install scriptlets are being run with a read-only filesystem.

**Comment by natoscott** [comment]
If the file /var/lib/pcp/pmns/root doesn't exist, the pmcd startup process (shell script that runs before pmcd binary) is supposed to create it even if .NeedRebuild doesn't exist.

Can someone run:  " sh -x /usr/share/pcp/lib/pmcd start "

and paste the output somewhere?  I'm particularly interested in seeing what happens in the _reboot_setup() function, which should execute the /var/lib/pcp/pmns/Rebuild script.

thanks!

**Comment by daniel-wtd** [comment]
@natoscott already done in my last post.

The full output could be found in this paste: https://pastebin.com/AkWBGx6q

**Comment by daniel-wtd** [comment]
@goodwinos I have done as suggested from you

```
sudo touch /var/lib/pcp/pmns/.NeedRebuild; sudo systemctl restart pmcd
```
The error is the same and the log states:

```
Log for pmcd on core00 started Fri May  1 06:21:50 2020

Warning: localhost only from PMCD_LOCAL=1 in environment
Error Parsing ASCII PMNS: Cannot open "/var/lib/pcp/pmns/root"
Error: pmLoadASCIINameSpace(DEFAULT, 1): No such file or directory
[Fri May  1 06:21:50] pmcd(105297) Error: pmcd not started due to errors!

Log finished Fri May  1 06:21:50 2020
```

**Comment by natoscott** [comment]
@daniel-wtd Aha - thanks!  The problem is evident at line 244 of the paste - it tests to see if the Rebuild script is executable, and skips rebuilding the namespace if it is not!

What filesystem permissions does /var/lib/pcp/pmns/Rebuild have there?  (ls -ld)

I'm guessing its going to have have execute permission stripped (for ostree / Silverblue reasons scripts aren't supposed to exist below /var?)

If that's the case, I wonder if fixing this is as simple as making that file a symlink and moving the actual script somewhere like /usr/share/pcp/lib/Rebuild so that the permissions get preserved?  It's line 161 and 166 of /usr/share/pcp/lib/pmcd that we need to influence.

**Comment by daniel-wtd** [comment]
@natoscott 

```
ls -ladZ /var/lib/pcp/pmns/
drwxr-xr-x. 2 root root system_u:object_r:pcp_var_lib_t:s0 4096 May  1 06:19 /var/lib/pcp/pmns/
```

If putting Scripts under /var is a good idea heavily depends on the "idea" of handling the FHS. ;)

**Comment by natoscott** [comment]
@daniel-wtd sorry - need the permissions on the actual script (Rebuild), not the directory itself.

**Comment by daniel-wtd** [comment]
@natoscott  The directory is empty, ;)

Only a ".NeedRebuild" file pops up.

**Comment by natoscott** [comment]
Oh boy.  :(  Anything at all below /var/lib/pcp/pmdas/linux/ by any chance?

**Comment by daniel-wtd** [comment]
Nope, empty. :(

**Comment by natoscott** [comment]
Does 'rpm -ql pcp' say there *should* be files there though?  Either way, I guess we're going to have to do a file migration from /var and into other places like /usr and /etc to resolve this (with back-compat symlinks for existing installations I guess).

**Comment by daniel-wtd** [comment]
Yup, rpm -ql states a lot, that should be located in /var/lib/pcp/pmdas and /var/lib/pcp/pmns.

**Comment by natoscott** [comment]
@daniel-wtd can you confirm - if/when you reboot, does that /var/lib/pcp/pmns/.NeedRebuild file you listed earlier there still exist?  thanks.

**Comment by daniel-wtd** [comment]
I have done a fresh installation of the x86 iso in a VM and rebooted:

```
# here you can find files
ls -la /var/lib/pcp/pmdas/

# this one is empty
ls -la /var/lib/pcp/pmns

# Starting the service the first time, crashing as expected
systemctl start pmcd
```

now doing as proposed, touching the file
```
touch /var/lib/pcp/pmns/.NeedRebuild
# Starting the service (2nd time)
systemctl start pmcd

# here you can find files
ls -la /var/lib/pcp/pmdas/

# this one contains the .NeedRebuild file
ls -la /var/lib/pcp/pmns
```

Afterwards, everything is unchanged and ":NeedRebuild" still exists.

```
reboot

# here you can find files
ls -la /var/lib/pcp/pmdas/

# this one contains the .NeedRebuild file
ls -la /var/lib/pcp/pmns
```

**Comment by natoscott** [comment]
@daniel-wtd well, that *is* good news.  :)

So, I think this is fixable via the process outlined below.  It needs to be cognizant of the fact we have an installed base of users with critical configuration files (like pmcd.conf) that reference binaries and libraries below /var - so simple models of just installing files to different locations in the next version of PCP is not going to fly.

1.  We need to fully audit everything we install below /var - from a quick (incomplete) audit, there's three broad categories - PMDAs, the PMNS, and configuration files.

2. We need to come up with an alternate, safe location for each of the files in these categories (better meeting the FHS in several cases) and $(INSTALL) to both the new location and the original (via symlink) in the PCP makefiles and spec files.

3.  We'll need a systemd tmpfiles setup for ostree that (re-)creates these symlinks on boot.  This should ideally be created during the build as each changed makefile creates the initial symlink.

I believe that'll do it - after that we'll have a working PCP installation on Silverblue.  Running this past @lberk this morning, and he agrees at least.  :)  Next step is finding a lucky volunteer to Make It So.

**Comment by daniel-wtd** [comment]
That sounds awesome. I am not the master at packaging. During this issue I had some glances at the provided scripts and tools, nothing that I am afraid of. 

If you are ok with some support from my side, I may support you. In the beginning you will need to reply to the most stupid questions, but supporting the project in one way or another would be awesome for me. 😍

**Comment by natoscott** [comment]
@daniel-wtd excellent, thanks - yep let's keep working together on this.  I have a bunch of other things on my plate, but I'll keep progressing this one in parallel as best I can.

After posting yesterday, I realised perhaps a better way of handling this situation will be to extend the install.sh script in the PCP git repo.  If we make it detect any regular file being installed into /var by the makefiles, it can a/ issue a warning and b/ acquire an 'alternative path' option for such files.  This could then be used to automate creation of the tmpfiles.d file for systemd.  Let me hack on that a little and get back to you.

**Comment by natoscott** [comment]
@daniel-wtd I'm about to push the makefile fixes and tmpfiles.d addition for PCP that address this issue - could you try things out and report back on how it all goes please?  (master branch)  Thanks!

**Comment by daniel-wtd** [comment]
Hi, I will give it a try this weekend. Thanks a lot.
